File: motor/imagenes.py
"""Procesamiento de imágenes y separadores (§5.6)."""

import re

SEPARADOR_XHTML = '<p class="hr centrado grande"><b>※ ・ ※ ・ ※</b></p>'

RE_IMG_PANDOC_P = re.compile(
    r'<p[^>]*>\s*<img\b[^>]*src="[^"]*?image0*(\d+)\.[a-zA-Z]+"[^>]*>\s*</p>',
    re.IGNORECASE,
)
RE_IMG_PANDOC = re.compile(
    r'<img\b[^>]*src="[^"]*?image0*(\d+)\.[a-zA-Z]+"[^>]*>',
    re.IGNORECASE,
)
RE_IMAGEN_TAG_P = re.compile(
    r'<p[^>]*>\s*\[IMAGEN\s*0*(\d+)\]\s*</p>',
    re.IGNORECASE,
)
RE_IMAGEN_TAG = re.compile(r'\[IMAGEN\s*0*(\d+)\]', re.IGNORECASE)
RE_SEPARADOR_P = re.compile(
    r'<p[^>]*>\s*\[(?:HR|SEPARADOR)\]\s*</p>',
    re.IGNORECASE,
)
RE_SEPARADOR = re.compile(r'\[(?:HR|SEPARADOR)\]', re.IGNORECASE)


def _reemplazo_imagen(match: re.Match) -> str:
    num_fmt = f'{int(match.group(1)):02d}'
    return (
        f'<hr class="sigil_split_marker" />\n'
        f'<figure class="dimg"><img src="../Images/{num_fmt}.jpg" alt=""/></figure>\n'
        f'<hr class="sigil_split_marker" />'
    )


def procesar_imagenes(html: str) -> tuple[str, int]:
    """Detecta imágenes nativas de pandoc y marcas [IMAGEN N] y las
    reemplaza por el formato ePub de Sigil. Devuelve el HTML y el conteo."""
    contador: list[int] = [0]

    def reemplazo(match: re.Match) -> str:
        contador[0] += 1
        return _reemplazo_imagen(match)

    html = RE_IMG_PANDOC_P.sub(reemplazo, html)
    html = RE_IMG_PANDOC.sub(reemplazo, html)
    html = RE_IMAGEN_TAG_P.sub(reemplazo, html)
    html = RE_IMAGEN_TAG.sub(reemplazo, html)
    return html, contador[0]


def procesar_separadores(html: str) -> tuple[str, int]:
    """Detecta marcas [HR] o [SEPARADOR] y las reemplaza por un
    separador XHTML estilizado. Devuelve el HTML y el conteo."""
    contador: list[int] = [0]

    def reemplazo(match: re.Match) -> str:
        contador[0] += 1
        return SEPARADOR_XHTML

    html = RE_SEPARADOR_P.sub(reemplazo, html)
    html = RE_SEPARADOR.sub(reemplazo, html)
    return html, contador[0]

File: app_flutter/lib/motor/imagenes.dart
final String separadorXhtml = '<p class="hr centrado grande"><b>※ ・ ※ ・ ※</b></p>';

final _reImgPandocP = RegExp(r'<p[^>]*>\s*<img\b[^>]*src="[^"]*?image0*(\d+)\.[a-zA-Z]+"[^>]*>\s*</p>', caseSensitive: false);
final _reImgPandoc = RegExp(r'<img\b[^>]*src="[^"]*?image0*(\d+)\.[a-zA-Z]+"[^>]*>', caseSensitive: false);
final _reImagenTagP = RegExp(r'<p[^>]*>\s*\[IMAGEN\s*0*(\d+)\]\s*</p>', caseSensitive: false);
final _reImagenTag = RegExp(r'\[IMAGEN\s*0*(\d+)\]', caseSensitive: false);
final _reSeparadorP = RegExp(r'<p[^>]*>\s*\[(?:HR|SEPARADOR)\]\s*</p>', caseSensitive: false);
final _reSeparador = RegExp(r'\[(?:HR|SEPARADOR)\]', caseSensitive: false);

String _reemplazoImagen(Match match) {
  int num = int.parse(match.group(1)!);
  String numFmt = num.toString().padLeft(2, '0');
  return '<hr class="sigil_split_marker" />\n'
      '<figure class="dimg"><img src="../Images/$numFmt.jpg" alt=""/></figure>\n'
      '<hr class="sigil_split_marker" />';
}

final _reHrDuplicados = RegExp(
  r'(<hr\s+class="sigil_split_marker"\s*/?>)(\s*(?:<!--.*?-->\s*)*<hr\s+class="sigil_split_marker"\s*/?>)+',
  caseSensitive: false,
);

String limpiarHrDuplicados(String html) {
  return html.replaceAll(_reHrDuplicados, r'$1');
}

final _reFiguraImg = RegExp(
  r'<figure[^>]*>\s*<img\b[^>]*src="[^"]*?(?:Images/|image0*)(\d+)\.[a-zA-Z]+"[^>]*>\s*</figure>',
  caseSensitive: false,
);

String? extraerPrimeraImagen(String html) {
  final match = _reFiguraImg.firstMatch(html);
  if (match != null) {
    int num = int.tryParse(match.group(1)!) ?? 1;
    return num.toString().padLeft(2, '0');
  }
  final matchTag = _reImagenTag.firstMatch(html);
  if (matchTag != null) {
    int num = int.tryParse(matchTag.group(1)!) ?? 1;
    return num.toString().padLeft(2, '0');
  }
  final matchPandoc = _reImgPandoc.firstMatch(html);
  if (matchPandoc != null) {
    int num = int.tryParse(matchPandoc.group(1)!) ?? 1;
    return num.toString().padLeft(2, '0');
  }
  return null;
}

typedef ProcesarResult = ({String html, int count});

ProcesarResult procesarImagenes(String html) {
  int contador = 0;
  String reemplazo(Match m) {
    contador++;
    return _reemplazoImagen(m);
  }
  html = html.replaceAllMapped(_reImgPandocP, reemplazo);
  html = html.replaceAllMapped(_reImgPandoc, reemplazo);
  html = html.replaceAllMapped(_reImagenTagP, reemplazo);
  html = html.replaceAllMapped(_reImagenTag, reemplazo);
  html = limpiarHrDuplicados(html);
  return (html: html, count: contador);
}

ProcesarResult procesarSeparadores(String html) {
  int contador = 0;
  String reemplazo(Match m) {
    contador++;
    return separadorXhtml;
  }
  html = html.replaceAllMapped(_reSeparadorP, reemplazo);
  html = html.replaceAllMapped(_reSeparador, reemplazo);
  return (html: html, count: contador);
}
